File: packages/discovery-provider/src/models/tracks/track.py
from sqlalchemy import (
    ARRAY,
    Boolean,
    Column,
    DateTime,
    Enum,
    Float,
    ForeignKey,
    Integer,
    PrimaryKeyConstraint,
    String,
    Text,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship, validates

from src.model_validator import ModelValidator
from src.models.base import Base
from src.models.model_utils import (
    RepresentableMixin,
    get_fields_to_validate,
    validate_field_helper,
)
from src.models.tracks.track_route import TrackRoute
from src.models.users.user import User


class Track(Base, RepresentableMixin):
    __tablename__ = "tracks"

    blockhash = Column(Text, nullable=False)
    blocknumber = Column(
        Integer, ForeignKey("blocks.number"), index=True, nullable=False
    )
    track_id = Column(Integer, primary_key=True, nullable=False)
    is_current = Column(Boolean, primary_key=True, nullable=False)
    is_delete = Column(Boolean, nullable=False)
    owner_id = Column(Integer, nullable=False, index=True)
    audio_upload_id = Column(String)
    preview_cid = Column(String, index=True)
    track_cid = Column(
        String, index=True
    )  # todo: after backfill, add nullable=False, both here and in a db migration
    orig_file_cid = Column(String, index=True)
    orig_filename = Column(Text)
    is_downloadable = Column(Boolean, nullable=False, server_default=text("false"))
    is_original_available = Column(
        Boolean, nullable=False, server_default=text("false")
    )
    title = Column(Text)
    duration = Column(Integer)
    preview_start_seconds = Column(Float)
    cover_art = Column(String)
    tags = Column(String)
    genre = Column(String)
    mood = Column(String)
    credits_splits = Column(String)
    create_date = Column(String)
    release_date = Column(DateTime)
    file_type = Column(String)
    metadata_multihash = Column(String)
    track_segments = Column(JSONB(), nullable=False)
    created_at = Column(DateTime, nullable=False, index=True)
    description = Column(String)
    isrc = Column(String)
    iswc = Column(String)
    license = Column(String)
    updated_at = Column(DateTime, nullable=False)
    cover_art_sizes = Column(String)
    is_scheduled_release = Column(Boolean, nullable=False, server_default=text("false"))
    is_unlisted = Column(Boolean, nullable=False, server_default=text("false"))
    field_visibility = Column(JSONB(True))
    route_id = Column(String)
    stem_of = Column(JSONB(True))
    remix_of = Column(JSONB(True))
    txhash = Column(
        String,
        primary_key=True,
        nullable=False,
        server_default=text("''::character varying"),
    )
    slot = Column(Integer)
    is_available = Column(Boolean, nullable=False, server_default=text("true"))
    allowed_api_keys = Column(ARRAY(String))
    is_stream_gated = Column(Boolean, nullable=False, server_default=text("false"))
    stream_conditions = Column(JSONB(True))
    is_download_gated = Column(Boolean, nullable=False, server_default=text("false"))
    download_conditions = Column(JSONB(True))
    is_playlist_upload = Column(Boolean, nullable=False, server_default=text("false"))
    playlists_containing_track = Column(
        ARRAY(Integer()), server_default="ARRAY[]::INTEGER[]"
    )
    playlists_previously_containing_track = Column(
        JSONB(True),
        nullable=False,
        server_default="json_build_object()",
    )
    ai_attribution_user_id = Column(Integer, nullable=True)
    placement_hosts = Column(String, nullable=True)
    pinned_comment_id = Column(Integer)

    # Audio analysis
    bpm = Column(Float)
    is_custom_bpm = Column(Boolean, server_default=text("false"))
    musical_key = Column(String)
    is_custom_musical_key = Column(Boolean, server_default=text("false"))
    audio_analysis_error_count = Column(
        Integer, nullable=False, server_default=text("0")
    )
    comments_disabled = Column(Boolean, nullable=False, server_default=text("false"))

    # From DDEX
    ddex_release_ids = Column(JSONB())
    ddex_app = Column(String)
    artists = Column(JSONB())
    resource_contributors = Column(JSONB())
    indirect_resource_contributors = Column(JSONB())
    rights_controller = Column(JSONB())
    copyright_line = Column(JSONB())
    producer_copyright_line = Column(JSONB())
    parental_warning_type = Column(String)
    cover_original_song_title = Column(String)
    cover_original_artist = Column(String)
    is_owned_by_user = Column(Boolean, nullable=False, server_default=text("false"))
    territory_codes = Column(ARRAY(String))
    no_ai_use = Column(Boolean, nullable=False, server_default=text("false"))
    parental_warning = Column(
        Enum(
            "explicit",
            "explicit_content_edited",
            "not_explicit",
            "no_advice_available",
            name="parental_warning_type",
        )
    )

    block1 = relationship(  # type: ignore
        "Block", primaryjoin="Track.blocknumber == Block.number"
    )

    _routes = relationship(  # type: ignore
        TrackRoute,
        primaryjoin="and_(\
            remote(Track.track_id) == foreign(TrackRoute.track_id),\
            TrackRoute.is_current)",
        lazy="joined",
        viewonly=True,
    )

    user = relationship(  # type: ignore
        User,
        primaryjoin="and_(\
            remote(Track.owner_id) == foreign(User.user_id),\
            User.is_current)",
        lazy="joined",
        viewonly=True,
    )

    @property
    def _slug(self):
        return self._routes[0].slug if self._routes else ""

    @property
    def permalink(self):
        if self.user and self.user[0].handle and self._slug:
            return f"/{self.user[0].handle}/{self._slug}"
        return ""

    PrimaryKeyConstraint(is_current, track_id, txhash)

    ModelValidator.init_model_schemas("Track")
    fields = get_fields_to_validate("Track")

    # unpacking args into @validates
    @validates(*fields)
    def validate_field(self, field, value):
        return validate_field_helper(field, value, "Track", getattr(Track, field).type)

    def get_attributes_dict(self):
        return {col.name: getattr(self, col.name) for col in self.__table__.columns}

File: packages/common/src/adapters/track.ts
import { full } from '@audius/sdk'
import type {
  CrossPlatformFile,
  Genre,
  Mood,
  TrackFilesMetadata,
  TrackMetadata
} from '@audius/sdk'
import camelcaseKeys from 'camelcase-keys'
import dayjs from 'dayjs'
import { omit, pick, mapValues } from 'lodash'
import snakecaseKeys from 'snakecase-keys'

import {
  Copyright,
  Id,
  OptionalId,
  RightsController,
  StemCategory,
  TrackSegment
} from '~/models'
import { StemTrackMetadata, UserTrackMetadata } from '~/models/Track'
import type { NativeFile, TrackMetadataForUpload } from '~/store/upload/types'
import { License, Maybe } from '~/utils'
import { decodeHashId } from '~/utils/hashIds'

import { accessConditionsFromSDK, accessConditionsToSDK } from './access'
import { resourceContributorFromSDK } from './attribution'
import { favoriteFromSDK } from './favorite'
import { coverArtSizesCIDsFromSDK } from './imageSize'
import { remixFromSDK } from './remix'
import { repostFromSDK } from './repost'
import { userMetadataFromSDK } from './user'
import { transformAndCleanList } from './utils'

export const trackSegmentFromSDK = ({
  duration,
  multihash
}: full.TrackSegment): TrackSegment => ({
  // Client code expects duration as a string
  duration: `${duration}`,
  multihash
})

export const userTrackMetadataFromSDK = (
  input: full.TrackFull | full.SearchTrackFull
): UserTrackMetadata | undefined => {
  const decodedTrackId = decodeHashId(input.id)
  const decodedOwnerId = decodeHashId(input.userId)
  const user = userMetadataFromSDK(input.user)
  if (!decodedTrackId || !decodedOwnerId || !user) {
    return undefined
  }

  const remixes = transformAndCleanList(input.remixOf?.tracks, remixFromSDK)

  const newTrack: UserTrackMetadata = {
    // Fields from API that are omitted in this model
    ...omit(snakecaseKeys(input), [
      'id',
      'user_id',
      'followee_favorites',
      'favorite_count',
      'is_streamable'
    ]),

    // Conversions
    track_id: decodedTrackId,
    owner_id: decodedOwnerId,
    // TODO: Remove this when api is fixed to return UTC dates
    release_date: input.releaseDate
      ? dayjs
          .utc(input.releaseDate)
          .local()
          // utc -> local
          .format('ddd MMM DD YYYY HH:mm:ss [GMT]ZZ')
      : null,

    // Nested Transformed Fields
    cover_art_cids: input.coverArtCids
      ? coverArtSizesCIDsFromSDK(input.coverArtCids)
      : null,
    download_conditions: input.downloadConditions
      ? accessConditionsFromSDK(input.downloadConditions)
      : null,
    field_visibility: snakecaseKeys(input.fieldVisibility),
    followee_saves: transformAndCleanList(
      input.followeeFavorites,
      favoriteFromSDK
    ),
    followee_reposts: transformAndCleanList(
      input.followeeReposts,
      repostFromSDK
    ),
    remix_of:
      remixes.length > 0
        ? {
            tracks: remixes
          }
        : null,
    stem_of: input.stemOf?.parentTrackId
      ? {
          category: input.stemOf.category as StemCategory,
          parent_track_id: input.stemOf.parentTrackId
        }
      : undefined,
    stream_conditions: input.streamConditions
      ? accessConditionsFromSDK(input.streamConditions)
      : null,
    track_segments: input.trackSegments.map(trackSegmentFromSDK),
    user,

    // Retypes
    license: (input.license as License) ?? null,

    // Nullable fields
    ai_attribution_user_id: input.aiAttributionUserId ?? null,
    allowed_api_keys: input.allowedApiKeys ?? null,
    artists: input.artists
      ? transformAndCleanList(input.artists, resourceContributorFromSDK)
      : null,
    audio_upload_id: input.audioUploadId ?? null,
    copyright_line: input.copyrightLine
      ? (snakecaseKeys(input.copyrightLine) as Copyright)
      : null,
    cover_art: input.coverArt ?? null,
    create_date: input.createDate ?? null,
    credits_splits: input.creditsSplits ?? null,
    ddex_app: input.ddexApp ?? null,
    ddex_release_ids: input.ddexReleaseIds ?? null,
    description: input.description ?? null,
    indirect_resource_contributors: input.indirectResourceContributors
      ? transformAndCleanList(
          input.indirectResourceContributors,
          resourceContributorFromSDK
        )
      : null,
    isrc: input.isrc ?? null,
    iswc: input.iswc ?? null,
    mood: input.mood ?? null,
    orig_file_cid: input.origFileCid ?? null,
    tags: input.tags ?? null,
    track_cid: input.trackCid ?? null,
    orig_filename: input.origFilename ?? null,
    parental_warning_type: input.parentalWarningType ?? null,
    preview_cid: input.previewCid ?? null,
    preview_start_seconds: input.previewStartSeconds ?? null,
    producer_copyright_line: input.producerCopyrightLine
      ? (snakecaseKeys(input.producerCopyrightLine) as Copyright)
      : null,
    resource_contributors: input.resourceContributors
      ? transformAndCleanList(
          input.resourceContributors,
          resourceContributorFromSDK
        )
      : null,
    rights_controller: input.rightsController
      ? (snakecaseKeys(input.rightsController) as RightsController)
      : null,
    save_count: input.favoriteCount,
    pinned_comment_id: input.pinnedCommentId ?? null,
    is_owned_by_user: input.isOwnedByUser,
    cover_original_artist: input.coverOriginalArtist ?? null,
    cover_original_song_title: input.coverOriginalSongTitle ?? null
  }

  return newTrack
}

export const stemTrackMetadataFromSDK = (
  input: full.StemFull
): StemTrackMetadata | undefined => {
  const [id, parentId, ownerId] = [input.id, input.parentId, input.userId].map(
    decodeHashId
  )
  if (!(id && parentId && ownerId)) return undefined

  return {
    blocknumber: input.blocknumber,
    is_delete: false,
    track_id: id,
    created_at: '',
    isrc: null,
    iswc: null,
    credits_splits: null,
    create_date: null,
    description: null,
    followee_reposts: [],
    followee_saves: [],
    genre: '',
    has_current_user_reposted: false,
    has_current_user_saved: false,
    license: null,
    mood: null,
    play_count: 0,
    owner_id: ownerId,
    release_date: null,
    repost_count: 0,
    save_count: 0,
    comment_count: 0,
    tags: null,
    title: '',
    track_segments: [],
    cover_art: null,
    cover_art_sizes: null,
    cover_art_cids: null,
    is_scheduled_release: false,
    is_unlisted: false,
    stem_of: {
      parent_track_id: parentId,
      category: input.category as StemCategory
    },
    artwork: {},
    remix_of: null,
    duration: 0,
    updated_at: '',
    permalink: '',
    is_available: true,
    is_stream_gated: false,
    stream_conditions: null,
    is_download_gated: false,
    download_conditions: null,
    access: { stream: true, download: true },
    track_cid: input.cid,
    orig_file_cid: '',
    orig_filename: input.origFilename,
    is_downloadable: true,
    is_original_available: false,
    is_playlist_upload: false,
    is_owned_by_user: false
  }
}

export const trackMetadataForUploadToSdk = (
  input: TrackMetadataForUpload
): TrackFilesMetadata => ({
  ...camelcaseKeys(
    pick(input, [
      'license',
      'isrc',
      'iswc',
      'is_unlisted',
      'is_premium',
      'premium_conditions',
      'is_stream_gated',
      'stream_conditions',
      'is_download_gated',
      'orig_filename',
      'is_downloadable',
      'is_original_available',
      'bpm',
      'is_custom_bpm',
      'musical_key',
      'is_custom_musical_key',
      'comments_disabled',
      'ddex_release_ids',
      'parental_warning_type'
    ])
  ),
  trackId: OptionalId.parse(input.track_id),
  title: input.title,
  description: input.description ?? undefined,
  mood: input.mood as Mood,
  tags: input.tags ?? undefined,
  genre: (input.genre as Genre) || undefined,
  releaseDate: input.release_date ? new Date(input.release_date) : undefined,
  previewStartSeconds: input.preview_start_seconds ?? undefined,
  previewCid: input.preview_cid ?? '',
  ddexApp: input.ddex_app ?? '',
  aiAttributionUserId: OptionalId.parse(input.ai_attribution_user_id),
  audioUploadId: input.audio_upload_id ?? undefined,
  duration: input.duration ?? undefined,
  trackCid: input.track_cid ?? '',
  origFileCid: input.orig_file_cid ?? '',
  fieldVisibility: input.field_visibility
    ? mapValues(
        camelcaseKeys(input.field_visibility),
        (value: Maybe<boolean>) => (value === null ? undefined : value)
      )
    : undefined,
  downloadConditions: input.download_conditions
    ? accessConditionsToSDK(input.download_conditions)
    : null,
  streamConditions: input.stream_conditions
    ? accessConditionsToSDK(input.stream_conditions)
    : null,
  remixOf: input.remix_of
    ? {
        tracks: input.remix_of.tracks.map((track) => ({
          parentTrackId: Id.parse(track.parent_track_id)
        }))
      }
    : undefined,
  stemOf: input.stem_of
    ? {
        category: input.stem_of.category,
        parentTrackId: Id.parse(input.stem_of.parent_track_id)
      }
    : undefined,
  copyrightLine: input.copyright_line
    ? camelcaseKeys(input.copyright_line)
    : undefined,
  producerCopyrightLine: input.producer_copyright_line
    ? camelcaseKeys(input.producer_copyright_line)
    : undefined,
  rightsController: input.rights_controller
    ? camelcaseKeys(input.rights_controller)
    : undefined,
  resourceContributors: input.resource_contributors
    ? input.resource_contributors.map((contributor) =>
        camelcaseKeys(contributor)
      )
    : undefined,
  indirectResourceContributors: input.indirect_resource_contributors
    ? input.indirect_resource_contributors.map((contributor) =>
        camelcaseKeys(contributor)
      )
    : undefined
})

export const fileToSdk = (
  file: Blob | NativeFile,
  name: string
): CrossPlatformFile => {
  // If we're in react-native
  if ('uri' in file) {
    return {
      buffer: Buffer.from(file.uri),
      name: file.name ?? name,
      type: file.type ?? undefined
    }
  }

  // If we're in browser (Blob)
  // If it's already a File, return as-is
  if (file instanceof File) {
    return file
  }

  // If it's a Blob, convert to File with a name
  return new File([file], name, { type: file.type })
}
